src/packet/udp.rs

```diff
@@ -26,54 +24,17 @@ pub struct Udp {
     payload: Vec<u8>,
 }
 
-/// Calculate the checksum for a packet built on IPv4
-pub fn ipv4_checksum(packet: &UdpPacket,
-                     ipv4_source: Ipv4Addr,
-                     ipv4_destination: Ipv4Addr,
-                     next_level_protocol: IpNextHeaderProtocol)
-    -> u16be {
-    let IpNextHeaderProtocol(next_level_protocol) = next_level_protocol;
-    let mut sum = 0u32;
-
-    // Checksum pseudo-header
-    // IPv4 source
-    let octets = ipv4_source.octets();
-    sum += (octets[0] as u32) << 8 | octets[1] as u32;
-    sum += (octets[2] as u32) << 8 | octets[3] as u32;
-
-    // IPv4 destination
-    let octets = ipv4_destination.octets();
-    sum += (octets[0] as u32) << 8 | octets[1] as u32;
-    sum += (octets[2] as u32) << 8 | octets[3] as u32;
-
-    // IPv4 Next level protocol
-    sum += next_level_protocol as u32;
-
-    // UDP Length
-    sum += (packet.packet()[4] as u32) << 8 | packet.packet()[5] as u32;
-
-    // Checksum UDP header/packet
-    let mut i = 0;
-    let len = packet.get_length() as usize;
-    while i < len && i + 1 < packet.packet().len() {
-        sum += (packet.packet()[i] as u32) << 8 | packet.packet()[i + 1] as u32;
-        i += 2;
-    }
-    // If the length is odd, make sure to checksum the final byte
-    if len & 1 != 0 && len <= packet.packet().len() {
-        sum += (packet.packet()[len - 1] as u32) << 8;
-    }
-    while sum >> 16 != 0 {
-        sum = (sum >> 16) + (sum & 0xFFFF);
-    }
-
-    !sum as u16
+fn checksum<T: PseudoHeader>(udp_packet: &UdpPacket, ip_packet: &T) -> u16 {
```

Comment:
Should return `u16be`, same goes for `rfc1071_checksum()`.

I also think it's a shame you have to pass both the IP packet and the UDP packet, but don't worry about that for now. I'll look at it when I'm working on #140.